### PR TITLE
add resume/reconnect to sdk with integration test

### PR DESCRIPTION
**Pull request by cnderrauber**
Add resume, reconnect to catch up other platform sdks. Add a integration test based on livekit docker image, need more test case to cover the go sdk

### CODE REVIEW

File: integration_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	host = "ws://localhost:7880"
+)
+
+var (
+	apiKey, apiSecret string
+)
+
+func TestMain(m *testing.M) {
+	keys := strings.Split(os.Getenv("LIVEKIT_KEYS"), ": ")
+	// keys := strings.Split("APIhdZda4TDAGxk: 4BA2qCorbmGnVZ9iMri7Sp0EEA7v2S4Oi8eyHuPxtxJ", ": ")
```
Comment (cnderrauber) [inline integration_test.go:28]: test key in the `mage test`, useful for debug when mage test failed

**Review by boks1971 — APPROVED** [review]
Nice work @cnderrauber !!!

File: engine.go [hunk 2]
```diff
@@ -315,3 +334,104 @@ func (e *RTCEngine) readDataPacket(msg webrtc.DataChannelMessage) (*livekit.Data
...
+		}
+
+		if e.OnDisconnected != nil {
+			e.OnDisconnected()
+		}
+	}()
+}
+
+func (e *RTCEngine) resumeConnection() error {
+	_, err := e.client.Join(e.url, e.token.Load(), e.connParams)
```
Comment (davidzhao) [inline engine.go:391]: I believe for resume, it needs to pass in `reconnect=1`, if it's not set, I don't think we would resume the session correctly.

File: integration_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+func TestResume(t *testing.T) {
+	rooms, err := createAgent(t.Name(), "publisher", "subscriber")
+	require.NoError(t, err)
+	pub, sub := rooms[0], rooms[1]
+	var reconnected atomic.Bool
+	pub.Callback.OnReconnected = func() {
+		reconnected.Store(true)
+	}
+	pub.Simulate(SimulateSignalReconnect)
+	require.Eventually(t, func() bool { return reconnected.Load() }, 500*time.Second, 100*time.Millisecond)
```
[the remark this answers is not in the document]
Comment (davidzhao) [reply integration_test.go:126]: WHAT? wow.. I had to write my own version of this.. great to know!

**Review by davidzhao — COMMENTED** [review]
fantastic work Jie!

Could you also add CI configuration so it'll run tests?

File: engine.go [hunk 4]
```diff
@@ -315,3 +334,104 @@ func (e *RTCEngine) readDataPacket(msg webrtc.DataChannelMessage) (*livekit.Data
...
+					logger.Error(err, "resume connection failed")
+					if !errors.Is(err, ErrSignalError) {
+						fullReconnect = true
+					}
+				} else {
+					return
+				}
+			}
+
+			delay := time.Duration(reconnectCount*reconnectCount) * initialReconnectInterval
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply engine.go:375]: want to keep same as other client(js)

File: engine.go [hunk 5]
```diff
@@ -315,3 +334,104 @@ func (e *RTCEngine) readDataPacket(msg webrtc.DataChannelMessage) (*livekit.Data
...
+		var fullReconnect bool
+		for ; reconnectCount < maxReconnectCount; reconnectCount++ {
+			if fullReconnect {
+				if reconnectCount == 0 && e.OnRestarting != nil {
+					e.OnRestarting()
+				}
+				if err := e.restartConnection(); err != nil {
+					logger.Error(err, "restart connection failed")
+				} else {
+					return
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply engine.go:359]: the return is under condition `err == nil`

**Review by davidzhao — APPROVED** [review]
looks good!

File: engine.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (boks1971) [reply engine.go:359]: Oops my bad, sorry @cnderrauber

**Pull request merged**